Paraview:  Use 'UNIX Makefiles' instead of ninja as the CMake generator for XL

quellyn opened the pull request:
Since the merging of #28650, `paraview` fails to build with the  IBM XL and Cray CCE compilers.

This issue has been [reported upstream to KitWare](https://gitlab.kitware.com/paraview/paraview/-/issues/21223#note_1140786) as a problem with `ninja` + `xl`. Since then we've encountered the same problems with `cce` builds.

This PR proposes that we fallback to the old generator ('Unix Makefiles') when building with `xl` and `cce`.

Including @alalazo and @vicentebolea, as they reviewed #28650, and also @EthanS94, as he confirmed the problem with `xl` and reported it to KitWare.

Thanks!

Commit:
hypre: Add releases 2.21.0 and 2.22.0

Commit:
Revert "hypre: Add releases 2.21.0 and 2.22.0"

This reverts commit 8921cdb3ac772794f82ff07b0aa816c204aca921.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Address external linkage failures in elfutils 0.185:
  https://bugs.gentoo.org/794601
  https://sourceware.org/pipermail/elfutils-devel/2021q2/003862.html
Encountered while building within a Spack environment.

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Revert "Address external linkage failures in elfutils 0.185:"

This reverts commit 76b93e4504defd856b62027371064a5e65a027a5.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into paraview_generator-for-cce-xl

Commit:
paraview: The ninja generator has problems with XL and CCE

See https://gitlab.kitware.com/paraview/paraview/-/issues/21223

kwryankrattiger commented:
I wish it could still use Ninja because building without it is not practical a lot of the time. But if it is blocking cce and xl compilers than I think this is a good solution.

quellyn commented:
> I wish it could still use Ninja because building without it is not practical a lot of the time. But if it is blocking cce and xl compilers than I think this is a good solution.

@kwryankrattiger, the patch is just falling back to 'Unix Makefiles' for `cce` and `xl` builds only -- the rest should build with `ninja`.

And, obviously there's a limit to the number of versions of these toolchains that I could test. These were:

- IBM XL C/C++ and Fortran for Linux (V16.1.1.7)
- IBM XL C/C++ and Fortran for Linux (V16.1.1.10)
- Cray CCE 12.0.3
- Cray CCE 13.0.1

If forcing all `cce` and `xl` compilers to use 'Unix Makefiles` is too heavy a hammer for every situation I could introduce a variant instead. Something like this:
```
variant('generator', default='Ninja', multi=False,
            values=('Ninja', 'Unix_Makefiles'),
            description='CMake generator.')
```

```
@property
def generator(self):
    return spec.variants['generator']
```

alalazo commented on var/spack/repos/builtin/packages/paraview/package.py:
```diff
@@ -220,6 +219,14 @@ class Paraview(CMakePackage, CudaPackage):
+    # https://gitlab.kitware.com/paraview/paraview/-/issues/21223
+    @property
+    def generator(self):
+        if self.spec.satisfies('%cce') or self.spec.satisfies('%xl') or self.spec.satisfies('%xl_r'):
```
There's one issue with this implementation: we have no dependency on `gmake` and we have an unconditional dependency on `ninja`. Adding a conditional dependency on `gmake` would be easy, but removing the one on `ninja` I think requires negation (which is currently not possible).

I wonder if turning `generator` into a sticky variant is possible in principle :thinking:

quellyn replied:
Hey, I have a possible solution. Let me test it with a couple of scenarios.

kwryankrattiger replied:
I am not a huge fan of adding more variants, especially since this doesn't actually change the binaries that are built.

It would be better to either understand why these compilers don't work with these compilers and fix that or put some of this logic into the `CMakePackage`.

quellyn replied:
> @quellyn
> 
> > the patch is just falling back to 'Unix Makefiles' for cce and xl builds only
> 
> I get that, I think that is a fine solution. I think we should explore further why are those compilers not working with Ninja. If there is not fix maybe this should go in the `CMakePackage`, then any other packages that need to use Ninja because builds take over a day without it should not have to duplicate/maintain this fix.

Definitely, the preferred solution would be to resolve the underlying issues with Ninja and the XL and CCE compilers. I think that will take more time and expertise than I have, however.

Really, this PR just seeks to restore basic functionality that was lost after the adoption of #28650 .

kwryankrattiger commented:
@quellyn 
> the patch is just falling back to 'Unix Makefiles' for cce and xl builds only

I get that, I think that is a fine solution. I think we should explore further why are those compilers not working with Ninja. If there is not fix maybe this should go in the `CMakePackage`, then any other packages that need to use Ninja because builds take over a day without it should not have to duplicate/maintain this fix.

Commit:
paraview: Add variant to allow choice of cmake generator.
This will be necessary until problems with cmake+ninja on XL and
CCE builds can be resolved.

See https://gitlab.kitware.com/paraview/paraview/-/issues/21223

quellyn commented:
> There's one issue with this implementation: we have no dependency on `gmake` and we have an unconditional dependency on `ninja`. Adding a conditional dependency on `gmake` would be easy, but removing the one on `ninja` I think requires negation (which is currently not possible).
> 
> I wonder if turning `generator` into a sticky variant is possible in principle 🤔

@alalazo, I'm having success with the variant approach. I was able to conditionally require either `gmake` or `ninja` using `when` like this:
```
depends_on('ninja', type='build', when='generator=ninja')
depends_on('gmake', type='build', when='generator=unix_makefiles')
```

I've tested this patch on the following:

- An XL build (`paraview%xl_r generator=unix_makefiles`
- A GCC build ( `paraview%gcc`)
 
The XL build correctly parses the variant value I pass it and sets the generator to 'Unix Makefiles' for cmake.
The GCC build correctly falls through to the default variant value and sets the generator to 'Ninja' for cmake.

kwryankrattiger commented:
I am still not thrilled about making this a variant since it isn't actually changing what is built.

Is there a link to the error you had with Ninja and those compilers? I was running on Perlmutter and Spock cray systems and didn't experience any issues related to Ninja there. I have a couple cycles I could dig some more.

quellyn commented:
> I am still not thrilled about making this a variant since it isn't actually changing what is built.
> 
> Is there a link to the error you had with Ninja and those compilers? I was running on Perlmutter and Spock cray systems and didn't experience any issues related to Ninja there. I have a couple cycles I could dig some more.

Please see this PR's description for a link to the issue filed with KitWare (for `xl` builds).

chuckatkins reviewed (changes requested):
I really don't want to make the generator a variant.  Changing the generator changes the way it builds but the installations from the two are no different; a variant would force spack to treat them as different.  Can it just be simplified like this?

```python
# Ideally this would have a when=`not (%xl or %cce)` but there's not a 
# good way to express that
depends_on('ninja', type='build')
...
def cmake_args(self):
    ...
    if not ('%xl' in spec or '%xl_r' in spec or '%cce' in spec):
        generator = 'Ninja'
    ...

quellyn commented:
> 

That was my first suggestion (e.g., no variant). Let me test that and I'll re-propose that going forward...

Commit:
paraview: ninja generator problems with XL/CCE
    By popular preference, abandon the idea of a special variant
    and select the generator based on compiler.

Commit:
Greg Becker suggested using the dedicated "generator" method to
pass the choice of makefile generator to cmake.

Commit:
paraview: The build errors I saw before with paraview%cce + ninja
have not reappeared in subsequent testing, so I'm dropping it from this
PR. If they re-occur I'll report the issue separately to KitWare.

chuckatkins commented on var/spack/repos/builtin/packages/paraview/package.py:
```diff
@@ -220,6 +219,14 @@ class Paraview(CMakePackage, CudaPackage):
+    @property
+    def generator(self):
+        # https://gitlab.kitware.com/paraview/paraview/-/issues/21223
+        if self.spec.satisfies('%xl') or self.spec.satisfies('%xl_r'):
```
So is it okay for cce?

quellyn replied:
Hi Chuck,

Sorry, I noted this in the last commit message but not in the ticket body. I can't reproduce my earlier failures with `cce`, so I've dropped it from the fix. If the problem resurfaces I'll document it in a separate issue, but for now I'd say we can limit this PR to just `xl` and `xl_r`.
